File: frontend/lib/course/duel_score.dart
/// Боевая математика игры «Ты против переводчика».
///
/// Копия правил из web/src/lib/duelScore.ts. Копия, а не общий код: у сайта и
/// приложения нет общего слоя логики, а расходиться правилам нельзя — бой на
/// телефоне и в браузере обязан считать одинаково.
///
/// Главное разделение. Полоски здоровья — это счёт и ничего кроме: выигранная
/// фраза снимает с соперника фиксированные двадцать, проигранная столько же с
/// вас, ничья задевает обоих поровну. Пять фраз по двадцать — ровно сто, то
/// есть чистая победа кладёт соперника, а любой другой расклад повторяет счёт
/// один в один.
///
/// Разгон от быстрого набора идёт в отдельную величину — «стиль». Он решает
/// ранг и признак критического удара, но до полосок не дотягивается. Иначе
/// зрелище начинает спорить с итогом: при быстром наборе две выигранные фразы
/// из пяти отдавали бы раунд человеку, который его проиграл.
library;

import 'dart:math' as math;

/// Здоровье на раунд. Раунд — пять предложений, потом полоски полны снова.
const int roundHp = 100;

/// Сколько секунд даётся на одно предложение.
const int sentenceSeconds = 40;

/// Что остаётся от разгона, когда время вышло.
const double clockDrain = 0.35;

/// Урон за фразу. Пять таких — ровно [roundHp], то есть чистый разгром.
const int winDamage = 20;

/// Ничья: качнулись оба и поровну.
const int tieDamage = 8;

/// Сила, начиная с которой удар считается критическим.
const double critPower = 82;

const double _powerFloor = 0.6;
const double _powerCeiling = 1.5;

enum DuelSide { hero, foe }

enum DuelWinner { user, translator, tie }

class Blow {
  /// Урон переводчику.
  final int toFoe;

  /// Урон вам.
  final int toHero;

  /// Очки стиля: разгон уходит сюда, а не в полоски.
  final int style;
  final bool crit;

  const Blow({
    required this.toFoe,
    required this.toHero,
    required this.style,
    required this.crit,
  });
}

/// Множитель разгона.
///
/// Пол не нулевой намеренно: перевод, обдуманный медленно, — всё равно
/// выигранный перевод, и оставлять его совсем без счёта значило бы наказывать
/// за раздумье в игре, которая учит языку.
double powerFactor(double power) {
  if (power.isNaN) return _powerFloor;
  final share = math.min(1.0, math.max(0.0, power / 100));
  return _powerFloor + (_powerCeiling - _powerFloor) * share;
}

/// Удар по итогам одного предложения.
Blow blowFor(DuelWinner winner, double power) => switch (winner) {
      DuelWinner.user => Blow(
          toFoe: winDamage,
          toHero: 0,
          style: (winDamage * powerFactor(power)).round(),
          crit: power >= critPower,
        ),
      DuelWinner.translator =>
        const Blow(toFoe: 0, toHero: winDamage, style: 0, crit: false),
      DuelWinner.tie => Blow(
          toFoe: tieDamage,
          toHero: tieDamage,
          style: (tieDamage * powerFactor(power)).round(),
          crit: false,
        ),
    };

/// Кто взял раунд по остатку здоровья.
DuelWinner roundOutcome(int heroHp, int foeHp) {
  if (heroHp == foeHp) return DuelWinner.tie;
  return heroHp > foeHp ? DuelWinner.user : DuelWinner.translator;
}

/// Ранг за раунд — по очкам стиля. Счёт он не заменяет.
String rankOf(int style) {
  if (style >= 130) return 'S';
  if (style >= 100) return 'A';
  if (style >= 60) return 'B';
  return 'C';
}

/// Ступень серии, 0..7 — под восемь файлов клика.
///
/// Логарифм, а не деление: первые удары серии слышны как рост, дальше разгон
/// замедляется, иначе к сотому символу звук упирается в потолок.
int comboStep(int streak) {
  if (streak <= 1) return 0;
  return math.min(7, (math.log(streak) / math.ln2).floor());
}

/// Отметки серии, на которых стоит крикнуть.
bool isComboMilestone(int streak) => streak > 0 && streak % 8 == 0;

File: web/src/lib/duelScore.ts
/**
 * Боевая математика игры «Ты против переводчика».
 *
 * Вынесена из сцены отдельно, потому что это её единственная проверяемая
 * часть. Тряску и искры проверяют глазами, а «полоски здоровья не могут
 * разойтись со счётом» — правило, которое молча перестанет работать, если его
 * не закрепить тестом. Оно и не работало: пока разгон усиливал урон, две
 * выигранные фразы из пяти при быстром наборе отдавали раунд человеку, хотя
 * по счёту он его проиграл.
 *
 * Отсюда главное разделение. Полоски здоровья — это счёт и ничего кроме:
 * выигранная фраза снимает с соперника фиксированные двадцать, проигранная
 * столько же с вас, ничья задевает обоих поровну. Пять фраз по двадцать — ровно
 * сто, то есть чистая победа кладёт соперника, а любой другой расклад
 * повторяет счёт один в один.
 *
 * Разгон, набранный быстрым набором, идёт в отдельную величину — «стиль». Он
 * решает ранг и то, будет ли удар критическим (звук, тряска, искры), но до
 * полосок не дотягивается. Так скоропечатание остаётся ставкой, а зрелище не
 * начинает спорить с итогом.
 */

/** Здоровье на раунд. Раунд — это пять предложений, потом полоски полны снова. */
export const ROUND_HP = 100;

/** Сколько секунд даётся на одно предложение. */
export const SENTENCE_SECONDS = 40;

/**
 * Что остаётся от разгона, когда время вышло.
 *
 * Здоровье часы не трогают: полоски принадлежат счёту. Цена просроченного
 * хода своя и вполне ощутимая — слабее стиль, ниже ранг, нет критов.
 */
export const CLOCK_DRAIN = 0.35;

/** Урон за фразу. Пять таких — ровно ROUND_HP, то есть чистый разгром. */
export const WIN_DAMAGE = 20;

/** Ничья: качнулись оба и поровну. */
export const TIE_DAMAGE = 8;

/** Сила, начиная с которой удар считается критическим. */
export const CRIT_POWER = 82;

/** Во сколько раз разгон усиливает стиль: от вялого набора до полного. */
const POWER_FLOOR = 0.6;
const POWER_CEILING = 1.5;

export type DuelWinner = 'user' | 'translator' | 'tie';

export interface Blow {
  /** Урон переводчику. */
  toFoe: number;
  /** Урон вам. */
  toHero: number;
  /** Очки стиля за этот удар: разгон уходит сюда, а не в полоски. */
  style: number;
  crit: boolean;
  /** Крупная надпись поверх удара. */
  label: string;
}

function clamp01(value: number): number {
  if (!Number.isFinite(value)) return 0;
  return Math.min(1, Math.max(0, value));
}

/**
 * Множитель разгона.
 *
 * Пол не нулевой намеренно. Перевод, обдуманный медленно и с паузами, — всё
 * равно выигранный перевод, и оставлять его совсем без счёта значило бы
 * наказывать за раздумье в игре, которая учит языку.
 */
export function powerFactor(power: number): number {
  return POWER_FLOOR + (POWER_CEILING - POWER_FLOOR) * clamp01(power / 100);
}

/** Удар по итогам одного предложения. */
export function blowFor(winner: DuelWinner, power: number): Blow {
  if (winner === 'user') {
    return {
      toFoe: WIN_DAMAGE,
      toHero: 0,
      style: Math.round(WIN_DAMAGE * powerFactor(power)),
      crit: power >= CRIT_POWER,
      label: 'ПОПАДАНИЕ',
    };
  }
  if (winner === 'translator') {
    return { toFoe: 0, toHero: WIN_DAMAGE, style: 0, crit: false, label: 'ПРОПУЩЕН' };
  }
  return {
    toFoe: TIE_DAMAGE,
    toHero: TIE_DAMAGE,
    style: Math.round(TIE_DAMAGE * powerFactor(power)),
    crit: false,
    label: 'РАЗМЕН',
  };
}

/** Кто взял раунд по остатку здоровья. */
export function roundOutcome(heroHp: number, foeHp: number): DuelWinner {
  if (heroHp === foeHp) return 'tie';
  return heroHp > foeHp ? 'user' : 'translator';
}

/**
 * Ранг за раунд — по очкам стиля.
 *
 * Счёт он не заменяет: кто выиграл предложения, уже решено. Ранг говорит
 * только о том, насколько уверенно: пять выигранных фраз без единой паузы
 * дают сто пятьдесят, те же пять с раздумьями — семьдесят пять.
 */
export function rankOf(style: number): string {
  if (style >= 130) return 'S';
  if (style >= 100) return 'A';
  if (style >= 60) return 'B';
  return 'C';
}

/**
 * Ступень серии, 0..7 — под восемь файлов клика (tools/build_duel_sounds.py).
 *
 * Логарифм, а не деление: первые удары серии должны быть слышны как рост, а
 * дальше разгон замедляется, иначе к сотому символу звук упирается в потолок
 * и перестаёт что-либо значить.
 */
export function comboStep(streak: number): number {
  if (streak <= 1) return 0;
  return Math.min(7, Math.floor(Math.log2(streak)));
}

/** Отметки серии, на которых стоит крикнуть. */
export function isComboMilestone(streak: number): boolean {
  return streak > 0 && streak % 8 === 0;
}
